File: app/lib/data/models/quiz.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'quiz.freezed.dart';
part 'quiz.g.dart';

/// One selectable answer in a quiz question (FEATURES_COMMUNITY_PLUS · Style Quiz).
@freezed
abstract class QuizOption with _$QuizOption {
  const factory QuizOption({
    required String key,
    required String label,
    @JsonKey(name: 'image_url') String? imageUrl,
  }) = _QuizOption;

  factory QuizOption.fromJson(Map<String, dynamic> json) =>
      _$QuizOptionFromJson(json);
}

@freezed
abstract class QuizQuestion with _$QuizQuestion {
  const factory QuizQuestion({
    required String id,
    required String prompt,
    @Default(<QuizOption>[]) List<QuizOption> options,
  }) = _QuizQuestion;

  factory QuizQuestion.fromJson(Map<String, dynamic> json) =>
      _$QuizQuestionFromJson(json);
}

@freezed
abstract class ActiveQuiz with _$ActiveQuiz {
  const factory ActiveQuiz({
    required String id,
    required String slug,
    required String title,
    String? description,
    @Default(<QuizQuestion>[]) List<QuizQuestion> questions,
  }) = _ActiveQuiz;

  factory ActiveQuiz.fromJson(Map<String, dynamic> json) =>
      _$ActiveQuizFromJson(json);
}

/// The computed "Style DNA" card.
@freezed
abstract class StyleResult with _$StyleResult {
  const factory StyleResult({
    required String title,
    @Default(<String>[]) List<String> keywords,
    @Default('') String description,
    @Default(<String>[]) List<String> palette,
  }) = _StyleResult;

  factory StyleResult.fromJson(Map<String, dynamic> json) =>
      _$StyleResultFromJson(json);
}

@freezed
abstract class QuizResult with _$QuizResult {
  const factory QuizResult({
    required String id,
    required StyleResult result,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _QuizResult;

  factory QuizResult.fromJson(Map<String, dynamic> json) =>
      _$QuizResultFromJson(json);
}

File: backend/app/models/quiz.py
"""Style Quiz models (FEATURES_COMMUNITY_PLUS · Style Quiz).

A short quiz whose result is a "Style DNA" card. Each option's `key` is a style
trait the backend tallies into the result; the result also feeds taste_signals.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class QuizOption(BaseModel):
    key: str
    label: str
    image_url: str | None = None


class QuizQuestion(BaseModel):
    id: str
    prompt: str
    options: list[QuizOption]


class ActiveQuiz(BaseModel):
    id: str
    slug: str
    title: str
    description: str | None = None
    questions: list[QuizQuestion]


class QuizSubmit(BaseModel):
    """The user's answers: {question_id: chosen option key}."""

    answers: dict[str, str] = Field(default_factory=dict)

    @model_validator(mode="after")
    def _require_answers(self) -> QuizSubmit:
        if not self.answers:
            raise ValueError("Answer at least one question.")
        return self


class StyleResult(BaseModel):
    """The computed Style DNA card."""

    title: str
    keywords: list[str]
    description: str
    palette: list[str] = Field(default_factory=list)


class QuizResult(BaseModel):
    id: str
    result: StyleResult
    created_at: datetime
